### ISSUE TITLE
Using Deno.core in embedded Deno

### ISSUE THREAD

**Issue by cd-work** [denoland/deno#18421]
Patch abd9610530 removed `Deno.core` and if I try to use `Deno.core.opAsync("whatever")` in my embedded Deno runtime now, I get the error that `Deno.core` is undefined.

I've tried to use `Deno[Deno.internal].core.opAsync` instead, since I saw that used in some places in Deno's codebase. While this is not undefined, it returns the following runtime error:

```
Uncaught TypeError: ops[name] is not a function
```

For an embedded Deno runtime, how is one supposed to use the `Deno.core` functionality?  My main goal is just the `opAsync` calls in this specific scenario.

A bit of additional context: The runtime constructed is through `MainWorker::bootstrap_from_options`. There is no startup snapshot (thought that it might be necessary?). The code is executed through `execute_main_module` and then `run_event_loop`. I can provide the full code if necessary.

**Comment by bartlomieju** [comment]
If you are using `deno_runtime` then you have to use `Deno[Deno.internal].core`.

The error you are getting is really and not great and we should improve it, and it basically means that the op you are trying to call is not registered with JavaScript side. Right now it is required to use a snapshot with `deno_runtime` - if you don't use it you should get a panic on startup.

I can provide more pointers, but I will need to see the actual code.

**Comment by bartlomieju** [comment]
> > Right now it is required to use a snapshot with deno_runtime - if you don't use it you should get a panic on startup.

> I'm using Default for the runtime, so it's possible I'm actually using a snapshot. I'm just not passing one myself.

Right, so you're using the default snapshot. All good here :+1:

> > The error you are getting is really and not great and we should improve it, and it basically means that the op you are trying to call is not registered with JavaScript side.

> So that means that the string passed to opAsync isn't a valid op? Did something change in the last couple of months with how these extensions need to be registered? Because I did not change anything about this so I would assume that just replacing Deno.core.opAsync with Deno[Deno.internal].core.opAsync should work?

> The relevant Deno code with the previous version (so in working condition) can be found here:
https://github.com/phylum-dev/cli/blob/main/cli/src/deno.rs#L34-L106

> The updates to the latest Deno version were minimal, so I didn't really change anything about the way this works other than switching to the Default impl for all the worker options. The registration of the ops is still identical.

Yes, we changed `deno_core` a bit - it now stores ops in the snapshot so they don't have to be registered on every startup (which incurred quite a bit of hit during startup). That means that extensions that are registered during startup, but not present during snapshotting need to "force" registration of ops. In your case it means you need to add `.force_op_registration()` on line 40 before `.build()`. Sorry for the breaking change, but we had to address the startup perf problem.

**Comment by cd-work** [comment]
> Sorry for the breaking change, but we had to address the startup perf problem.

Absolutely no problem, I'm personally always happy about performance improvements. As long as you're so quick to help out with issues, it's a welcome change.

That said I did check for a changelog somewhere and did not find it. I'm not sure if it would have been worth it for this issue alone, but just letting you know that I would have checked a changelog before opening this issue had I found one (which could have potentially solved this).
